TOT-80 : [FEAT] 내 계획 목록 검색 및 조회 API 구현

honee8583 opened the pull request:
## PR전 체크리스트
> 1. PR 및 Commit title을 형식에 맞게 작성했나요(e.g. TOT-123 : [TYPE]) O
> 2. Reviewers가 명확하게 지정됐나요? O
> 3. Assignees가 명확하게 지정됐나요? O


## 📝 작업 내용
- 검색요청값에 따른 내 계획 목록 조회 동적 조회 메서드 작성
  - 검색 요청 타입에 제목타입이 들어가거나 값이 들어오지 않을 경우 제목 검색 로직 실행
  - 검색 요청 타입에 지역타입이 들어갈경우 지역 검색 로직 실행
  - 기본값으로 다가오는 여행으로 및 최근 6개월안에 작성된 계획 목록을 조회
- 지역검색의 경우 지역이름에 키워드가 포함된 Region을 기준으로 PlanRegion을 조인후 PlanRegion의 planId를 먼저 페이징처리하여 조회
  (조회한 planId들과 연관된 Region의 name들을 조회하여 파싱)

### 스크린샷


## 💬 리뷰 요구사항
- 제가 작성한 검색 로직은 검색요청값에 해당되는 planId값들을 먼저 페이징 처리하여 조회한 후 각 planId에 연관된 Region의 이름들을 불러오는 코드입니다. 따라서 한번의 요청으로 planId 조회 로직, 카운트 쿼리 조회, 그리고 각 planId에 해당하는 Plan과 연관된 Region 조회로 최대 10번의 요청을 하게 됩니다(한 페이지당 최대 10개의 데이터를 포함하므로). 또한, 지역 검색의 경우 PlanRegion을 기준으로 Region과 조인하여 지역 이름이 포함된 Region과 연관된 PlanRegion의 planId값을 페이징처리하여 먼저 조회하고, 조회한 각 planId를 기준으로 Plan을 조회하게 됩니다. 이보다 더 나은 방식과 요청횟수를 줄일 수 있는 방법이 있다면 조언해주시면 감사하겠습니다!
- 이 외에 놓친부분이 있다면 말씀해주시면 감사하겠습니다!


## 참고자료

honee8583 committed:
TOT-126 : [FIX] PlanRegion 순환 참조 해결

honee8583 committed:
TOT-80 : [FEAT] Plan 검색 관련 DTO 클래스 생성

- PlanSearchValue : 검색 관련 상수 모음.
- PlanSearchRequestDto : 검색 타입(T/R), 키워드, 다가오는여행/지난여행 여부, 최근 6개월/년도 포함.
- PlanListResponseDto : 검색 목록 조회 결과 DTO
- PageResponseDto : 페이징 정보를 포함한 결과 DTO

honee8583 committed:
TOT-80 : [FEAT] PlanRepository 확장

honee8583 committed:
TOT-80 : [FEAT] 내 계획 목록 조회 서비스 메서드 작성

honee8583 committed:
TOT-80 : [FEAT] 내 계획 목록 동적 조회 로직 작성

- searchPlans() : 검색 요청값이 제목검색일경우 제목으로, 지역검색일경우 지역으로 검색.
- searchPlansByTitle() : 제목 검색 메서드. 유저아이디와 동일, 제목에 키워드를 포함, 다가오는여행/지난여행 여부, 작성일에 부합하고 요청한 페이지에 맞는 planId 를 조회하는 쿼리, 이에 해당하는 전체 데이터 카운트 쿼리 수행후 PageImpl 객체 반환.
- searchPlansByRegion() : 지역 검색 메서드. 유저아이디와 동일, 연관된 지역에 키워드에 해당하는 지역명을 포함, 다가오는여행/지난여행 여부, 작성일에 부합하고 요청한 페이지에 맞는 planId 를 조회하는 쿼리와 똑같이 카운트 쿼리 수행후 PageImpl 객체 반환.
- search() : 각 검색 메서드에서 조회한 planId와 연관된 PlanRegion, Region 테이블을 조인해 계획정보와 지역정보를 조회해 List<PlanListResponseDto> 로 반환.
- userEq() : Plan의 User엔티티의 id 값과 요청한 userId와 동일여부.
- planRegionUserIdEq() : PlanRegion의 Plan의 User 엔티티의 id 값과 요청한 userId와 동일여부.
- planDateGoeOrLt() : 다가오는 여행을 검색했을 경우 계획 시작날짜가 현재날짜보다 크거나 같은지에 대한 여부, 지난 여행을 검색했을 경우 계획 종료날짜가 현재날짜보다 작은지에 대한 여부.
- planCreatedInYear() : 작성일이 최근 6개월인 계획을 검색 혹은 요청한 년도에 작성된 계획 여부.
- containsKeyword() : 검색 요청 값에 타입이 "T"이거나 존재하지 않을경우 제목 검색, 타입이 "R"일경우 지역검색.
- titleContainsKeyword() : 계획의 제목에 키워드 포함 여부.
- regionNameContainsKeyword() : 지역의 이름에 키워드 포함 여부.
- isSearchByTitle() : 제목검색, 지역검색 여부 판단.

honee8583 committed:
TOT-80 : [FEAT] 내 계획 목록 검색 및 조회 컨트롤러 메서드 작성

honee8583 committed:
TOT-80 : [TEST] 내 계획 목록 검색 Querydsl 테스트 코드 작성

- 각 테스트 코드 전에 데이터를 저장할시 id값이 초기화 되지 않으므로 @BeforeEach가 실행되기 전에 auto_increment를 1로 초기화하는 sql 작성.

honee8583 committed:
TOT-80 : [TEST] 내 계획 목록 검색 및 조회 서비스 메서드 테스트 코드 작성

honee8583 committed:
TOT-80 : [TEST] 내 계획 목록 검색 및 조회 컨트롤러 메서드 테스트 코드 작성

honee8583 committed:
TOT-80 : [REFACTOR] PageResponseDto를 common패키지로 이동

MisaSohee commented on backend/src/main/java/com/triportreat/backend/common/response/PageResponseDto.java:
현재 페이지에 속한 페이지 블록의 마지막 페이지 를 계산할 때(calculateEndPage 메서드), 마지막 아이템의 인덱스번호를 비교대상에 두신 이유가 무엇인가요?? 리뷰가 몇 번째에 위치하는지랑 페이지 수랑 관련이 없다는 생각이 들어서 여쭤봅니다!

totalPages랑 startPage + blockSize - 1 랑 비교하는 방식으로 가는건 어떤가요??

honee8583 replied:
잘 이해가 안가는것 같네요ㅠ 계산방식을 조금 바꿔놓은 상태인데 혹시 이 방식을 말씀하시는걸까요? 아니라면 시간되실때 구글미트로 설명 조금 부탁드릴게요!

MisaSohee reviewed (approved):
고생많으셨습니다!

snowcrab382 commented:
다시 한 번 확인차 여쭤봅니다! PR에 대한 검토는 그때 말씀드린 GROUP_CONCAT과 서브쿼리를 활용한 로직으로 수정하신 이후에 진행하면 되는 게 맞을까요??

honee8583 committed:
TOT-80 : [TEST] auto_increment 초기화 구문 추가

- teardown.sql 삭제
- autoIncrementReset() : Plan, Plan_Region, User 테이블 auto_increment 1로 초기화. (h2 데이터베이스는 'USER'를 예약어로 인식하므로 이스케이프 문자 추가)

honee8583 committed:
TOT-80 : [FIX] 컨트롤러에 사용자 아이디 파라미터 추가

honee8583 commented:
이제 PR 검토 해주셔도 될것 같아요. 아무래도 다른 기능들 먼저 마무리를 하고 리팩토링으로 진행할것 같습니다!

honee8583 committed:
TOT-80 : [REFACTOR] 페이징 응답 필드 변경

- first, last 필드 제거
- 블록사이즈 5로 변경
- 현재 화면의 시작페이지번호, 마지막 페이지번호 계산방법 수정

snowcrab382 reviewed (approved):
고생하셨습니다! 로그가 곳곳에 찍혀있는데 더 이상 안쓰신다면 제외시키는게 좋을 것 같습니다.

honee8583 committed:
TOT-80 : [REFACTOR] PlanRepository 각 테스트 코드 실행전 PersistenceContext 초기화

- 불필요한 로그 삭제

honee8583 committed:
TOT-80 : [REFACTOR] 내 계획 목록 조회 API 주소 변경

honee8583 committed:
TOT-80 : [REFACTOR] 계획목록조회 서비스 메서드 리팩토링

honee8583 committed:
TOT-80 : [FEAT] DateUtil 클래스 생성

honee8583 committed:
TOT-80 : [REFACTOR] String 타입의 작성일 필드 추가

- LocalDateTime 타입의 작성일은 @JsonIgnore 설정

honee8583 committed:
TOT-80 : [REFACTOR] 작성일 변환 로직 작성

honee8583 committed:
TOT-80 : [TEST] 테스트 코드 수정
